### PR TITLE
minor quic improvements

### PR DESCRIPTION
**Pull request by geoffkizer**
Add a test, fix an assert, etc.

### CODE REVIEW

File: src/libraries/System.Net.Quic/tests/FunctionalTests/QuicStreamTests.cs [hunk 1]
```diff
@@ -228,6 +228,84 @@ public async Task GetStreamIdWithoutStartWorks()
...
+                    }
+
+                    for (int i = 0; i < NumberOfWrites; i++)
+                    {
+                        await stream.WriteAsync(data);
+                    }
+
+                    await stream.WriteAsync(Memory<byte>.Empty, endStream: true);
+
+                    await connection.CloseAsync(errorCode: 0);
```
Comment (scalablecory) [inline src/libraries/System.Net.Quic/tests/FunctionalTests/QuicStreamTests.cs:268]: You need to wait for stream shutdown before closing the connection, or the stream data is not guaranteed to get there.

Comment (geoffkizer) [reply src/libraries/System.Net.Quic/tests/FunctionalTests/QuicStreamTests.cs:268]: Thanks. Presumably if I Dispose the stream, I get the guarantee?

Comment (geoffkizer) [reply src/libraries/System.Net.Quic/tests/FunctionalTests/QuicStreamTests.cs:268]: Alternatively, if I just remove the call to connection.CloseAsync and let it be disposed by the using statement, will that guarantee that the data is sent as well?

I'm trying to understand how customers are expected to use this.

Comment (scalablecory) [reply src/libraries/System.Net.Quic/tests/FunctionalTests/QuicStreamTests.cs:268]: Not in current implementation. My proposal in https://github.com/dotnet/runtime/issues/756#issuecomment-697979873 would make that the case.

Comment (geoffkizer) [reply src/libraries/System.Net.Quic/tests/FunctionalTests/QuicStreamTests.cs:268]: ah, right.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/Mock/MockStream.cs [hunk 2]
```diff
@@ -222,10 +222,16 @@ internal sealed class StreamState
             public long _outboundErrorCode;
             public long _inboundErrorCode;
 
+            private const int InitialBufferSize =
+#if DEBUG
+                10;
+#else
+                4096;
+#endif
```
Comment (stephentoub) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/Mock/MockStream.cs:230]: Can we similarly set the max buffer size to something very small?

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/Mock/StreamBuffer.cs [hunk 3]
```diff
@@ -340,7 +340,10 @@ private void CancelWaiter()
 
             public void Reset()
             {
-                Debug.Assert(_hasWaiter == 0);
+                if (_hasWaiter != 0)
+                {
+                    throw new InvalidOperationException("Concurrent use is not supported");
```
Comment (stephentoub) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/Mock/StreamBuffer.cs:345]: I'd suggest Debug.Fail'ing as well.  That'll fail fast, be less likely to be eaten, provide a better debugging experience (e.g. crash dump), etc.

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/Mock/MockStream.cs (hunk 2 above)
Comment (geoffkizer) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/Mock/MockStream.cs:230]: Sure. I made it 64.

Comment (geoffkizer) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/Mock/MockStream.cs:230]: Making it so small is making the "large data" test take forever to run.

Even at 4096 bytes, it still takes several seconds to run.

I'm going to set this to 4096 and reduce the size in the test to 16MB (from 64MB).

**Pull request merged**